### PR TITLE
Add branding event handler implementation.

### PR DESCRIPTION
**Pull request by UdeshAthukorala**
## Approach
- Introduce `PRE_ADD_BRANDING_PREFERENCE` event for trigger before adding a new branding preference.
- Introduce `PRE_UPDATE_BRANDING_PREFERENCE` event for trigger before updating a branding preference

### CODE REVIEW

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java [hunk 1]
```diff
@@ -48,4 +50,24 @@ public void setConfigurationManager(ConfigurationManager configurationManager) {
 
         this.configurationManager = configurationManager;
     }
+
+    /**
+     * Return identity event service.
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java:55]: Get identity event service reference.

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java [hunk 2]
```diff
@@ -307,6 +324,29 @@ private String getResourceName(String type, String name, String locale) {
...
     }
 
+    /**
+     * Trigger branding preference validation.
+     *
+     * @param brandingPreference Branding Preference.
+     * @param tenantDomain       Tenant domain.
+     */
+    private void triggerBrandingPreferenceValidation(BrandingPreference brandingPreference, String tenantDomain)
+            throws BrandingPreferenceMgtException {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java:334]: triggerBrandingPreferenceEvents

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java [hunk 3]
```diff
@@ -321,5 +361,9 @@ private ConfigurationManager getConfigurationManager() {
 
         return BrandingPreferenceManagerComponentDataHolder.getInstance().getConfigurationManager();
     }
-}
 
+    private IdentityEventService getIdentityEventService() {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java:365]: Don't see much of a usage of this method

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java (hunk 1 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java:55]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/3/commits/6c642e68405f6962618070bef2176e84b66cf5c0

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java (hunk 2 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java:334]: Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/commit/6c642e68405f6962618070bef2176e84b66cf5c0

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java [hunk 4]
```diff
@@ -307,6 +329,59 @@ private String getResourceName(String type, String name, String locale) {
...
+     * @param brandingPreference Branding Preference.
+     * @param tenantDomain       Tenant domain.
+     */
+    private void triggerPreAddBrandingPreferenceEvents(BrandingPreference brandingPreference, String tenantDomain)
+            throws BrandingPreferenceMgtException {
+
+        Map<String, Object> eventProperties = new HashMap<>();
+        eventProperties.put(BRANDING_PREFERENCE, brandingPreference);
+        eventProperties.put(TENANT_DOMAIN, tenantDomain);
+        Event brandingPreferenceEvent = new Event(PRE_ADD_BRANDING_PREFERENCE, eventProperties);
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java:344]: Shall we add this event as well to the PR description?

Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/BrandingPreferenceManagerImpl.java:344]: Updated

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java (hunk 1 above)
Comment (madurangasiriwardena) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java:55]: I don't see this updated yet.

**Review by madurangasiriwardena — APPROVED** [review]

File: components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java (hunk 1 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.branding.preference.management.core/src/main/java/org/wso2/carbon/identity/branding/preference/management/core/internal/BrandingPreferenceManagerComponentDataHolder.java]: Adressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/3/commits/35cccfb9edb1ac51496089e3de3c3c7ba2745443

**Pull request merged**